src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -541,7 +542,22 @@ trait Applications extends Compatibility { self: Typer =>
 
     def realApply(implicit ctx: Context): Tree = track("realApply") {
       var proto = new FunProto(tree.args, IgnoredProto(pt), this)(argCtx(tree))
-      val fun1 = typedExpr(tree.fun, proto)
+
+      /* In the default case we type using Mode.IgnoreNextDynamic because fun1 might be a Select(foo, bar) where foo is of
+       * type scala.Dynamic and bar is not one of its members. In which case the following transformation will be tried:
+       *   foo.bar(baz0, baz1, ...) ~~> foo.applyDynamic(bar)(baz0, baz1, ...)
+       * Mode.IgnoreNextDynamic will ensure that Apply(Select(foo, bar), baz) is not transformed
+       * into foo.selectDynamic(bar).apply(args).
+       * We remove Mode.IgnoreNextDynamic if the name of the called method if is one of the scala.Dynamic methods or
+       * if this node was generated by an assign (see typedAssign).
+       * See: Dynamic.scala
+       */
+      def generatedByTypedAssign(name: Name) = name == nme.update &&  ctx.mode.is(Mode.IgnoreNextDynamic)
+      val ctx1 = tree.fun match {
+        case Select(_, name) if isDynamicMethod(name) || generatedByTypedAssign(name) => ctx.retractMode(Mode.IgnoreNextDynamic)
+        case _                                                                        => ctx.addMode(Mode.IgnoreNextDynamic)
```

Comment:
That's a bit heavy (both computationally, addMode is not super-cheap, and in terms for complexity). Is there another way to achieve this?